### DISCUSSION TITLE
Aiortc Behind NAT (no publicly addressable IP)

### DISCUSSION THREAD

**Discussion by savantspecter** [aiortc/aiortc#763] [Q&A]
TLDR: Is using aiortc from a server behind NAT with no public IP (using a TURN server) possible?

I read on an old thread (I think 2018) that aiortc when running as a server expects to have a publicly addressable IP. Is this still the case? This would imply that aiortc doesn't expect to use a TURN server. Would port forwarding a port change things?

I am trying to get the server example working on a windows desktop computer behind NAT.  If the client is local everything works. Additionally, if I spin up a linux/windows AWS instance with a public addressable IP, I can get the code to work as well - client uses the TURN server, aiortc server set to not use it.

However when run on the desktop I get an issue that I can't figure out. A few notes
1) similar to this thread (https://github.com/aiortc/aiortc/issues/131) -- The local description as an answer from the aiortc server (as related address from the TURN server) is the local IP 192.168... instead of having a related address equal to the network's public IP. When I uses 2 browsers (and dont use aiortc) I get the related address is the network's public IP. So I assume there is an issue there.

2) An initial connection is established but then aiortc on verbose mode notes a DTLS handshake failure. I dont know if this is a result of (1) or completely separate.

Note that: the aiohttp server part is accessible through reverse tunneling and so can already be accessed publicly. But I'm not sure this is relevant for the webrtc part.

Any thoughts would be really helpful. I've spent a significant amount of time trying to understand and debug this as I would really like to use this tech. I appreciate the effort that went into making this package and thanks for any help!

------------------------------------------

Running the script from here (https://github.com/aiortc/aiortc/issues/252) but inputting my TURN server I get this readout:
Notably the relatedAddress for the TURN server (last candidate) is the local IP not the public IP

[RTCIceCandidate(
component=1,
foundation='34e4f7e86cce16212cee5d77eee12e6c',
ip='192.168.86.56',
port=63640,
priority=2130706431,
protocol='udp',
type='host',
relatedAddress=None,
relatedPort=None,
sdpMid=None,
sdpMLineIndex=None,
tcpType=None),

RTCIceCandidate(
component=1,
foundation='e03b9305f1982b0795b015b743b7e2ae',
ip='fd8f:710a:6527:0:94e6:6679:2640:598e',
port=63641,
priority=2130706431,
protocol='udp',
type='host',
relatedAddress=None,
relatedPort=None,
sdpMid=None,
sdpMLineIndex=None,
tcpType=None),

RTCIceCandidate(
component=1,
foundation='b91e4e2817ea9a19c7fe05592cc6faf5',
ip='fd8f:710a:6527:0:4459:63b4:b707:c51f',
port=63642,
priority=2130706431,
protocol='udp',
type='host',
relatedAddress=None,
relatedPort=None,
sdpMid=None,
sdpMLineIndex=None,
tcpType=None),

RTCIceCandidate(
component=1,
foundation='81f75aaa651882210268504113750c71',
ip='172.31.93.153',
port=56003,
priority=16777215,
protocol='udp',
type='relay',
relatedAddress='192.168.86.56',
relatedPort=54621,
sdpMid=None,
sdpMLineIndex=None,
tcpType=None)]

**Comment by mfalberts** [comment]
Did you find a solution to this issue?  I too have been struggling with this for days (weeks actually off and on).  Would really appreciate any additional information you may have found out. Thanks!

**Reply by savantspecter** [reply]
Short answer no. But it's been a while, so I dont remember all the exact details

1. I only had issues running the python server being the firewall. The versions of webrtc used by the browser worked fine. So as a temporary workaround I was spinning up browser versions and controlling them with a server. Needless to say it was a bit of a mess, but it was working for my testing. Notably this means that it should have been possible, I just couldn't make it work when I was controlling it with a (python server)
2. On a whim, significantly after I posted here, I tried to ask chatgpt what it could find (as chatgpt wasn't as good back when I originally posted). It did seem like chatgpt had a way to specifically pass your own self made ICE candidate. As I knew what my path would be, this would have worked for me in theory. I dont know where chatgpt found that info on creating your own ice candidate though, because through all of my research I couldn't find any documentation on creating your own. So I can't say for sure chatgpt was hallucinating

Hope it helps!
